

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>2.12. Loading Coexpression analyzes &mdash; machado 0.1.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=01f34227"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="2.13. Loading VCF files" href="load_vcf.html" />
    <link rel="prev" title="2.11. Loading RNA-seq data" href="load_rnaseq.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            machado
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">1. Installation</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="dataload.html">2. Data loading</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="load_ontologies.html">2.1. Ontologies</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_taxonomy.html">2.2. Taxonomy</a></li>
<li class="toctree-l2"><a class="reference internal" href="insert_organism.html">2.3. Organism</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_publication.html">2.4. Publication</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_fasta.html">2.5. FASTA</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_gff.html">2.6. GFF</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_feature_annotation.html">2.7. Feature additional info</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_blast.html">2.8. Blast</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_interproscan.html">2.9. InterproScan</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_orthomcl.html">2.10. OrthoMCL</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_rnaseq.html">2.11. RNA-seq</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">2.12. Coexpression</a></li>
<li class="toctree-l2"><a class="reference internal" href="load_vcf.html">2.13. VCF</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="visualization.html">3. Visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="diagrams.html">4. Diagrams</a></li>
<li class="toctree-l1"><a class="reference internal" href="models.html">5. Models</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">machado</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="dataload.html"><span class="section-number">2. </span>Data loading</a></li>
      <li class="breadcrumb-item active"><span class="section-number">2.12. </span>Loading Coexpression analyzes</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/load_coexpression.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="loading-coexpression-analyzes">
<h1><span class="section-number">2.12. </span>Loading Coexpression analyzes<a class="headerlink" href="#loading-coexpression-analyzes" title="Link to this heading"></a></h1>
<section id="load-coexpression-correlated-pairs-of-features">
<h2><span class="section-number">2.12.1. </span>Load Coexpression correlated pairs of features<a class="headerlink" href="#load-coexpression-correlated-pairs-of-features" title="Link to this heading"></a></h2>
<p>Coexpression analyzes usually generate correlation statistics regarding gene sets in a pairwise manner.
For example, LSTrAP can generate a table with the Pearson correlation coefficient for every pair of genes in
a RNA-seq experiment.</p>
<p>To load such a table into Machado, the file must be headless and tab separated, with the two first columns
containing the correlated pair of IDs for the genes/features and the third column must contain the correlation
coefficient among them. In the case of the output of the LSTrAP software, the coefficient is subtracted by 0.7
for normalization sakes.</p>
<p>A one line sample from such a table is showed below:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>AT2G44195.1.TAIR10<span class="w">      </span>AT1G30080.1.TAIR10<span class="w">      </span><span class="m">0</span>.18189286870895194
AT2G44195.1.TAIR10<span class="w">      </span>AT5G24750.1.TAIR10<span class="w">      </span><span class="m">0</span>.1715779378273995
</pre></div>
</div>
<p>Note: The feature pairs from columns 1 and 2 need to be loaded previously.</p>
<p>To load such a table, one can run the command below:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>manage.py<span class="w"> </span>load_coexpression_pairs<span class="w"> </span>--organism<span class="w"> </span><span class="s1">&#39;Arabidopsis thaliana&#39;</span><span class="w"> </span>--file<span class="w"> </span>pcc.mcl.txt
</pre></div>
</div>
<ul class="simple">
<li><p>Loading this file can be faster if you increase the number of threads (–cpu).</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>manage.py<span class="w"> </span>load_coexpression_pairs<span class="w"> </span>--help
</pre></div>
</div>
<table class="docutils align-default">
<tbody>
<tr class="row-odd"><td><p>–file</p></td>
<td><p>‘pcc.mcl.txt’ File *</p></td>
</tr>
<tr class="row-even"><td><p>–organism</p></td>
<td><p>Species name (eg. Homo sapiens, Mus musculus) *</p></td>
</tr>
<tr class="row-odd"><td><p>–soterm</p></td>
<td><p>‘mRNA’ sequence ontology term <em>default</em></p></td>
</tr>
<tr class="row-even"><td><p>–cpu</p></td>
<td><p>Number of threads</p></td>
</tr>
</tbody>
</table>
<p>* example output file from LSTrAP software.</p>
</section>
<section id="remove-coexpression-pairs">
<h2><span class="section-number">2.12.2. </span>Remove coexpression pairs<a class="headerlink" href="#remove-coexpression-pairs" title="Link to this heading"></a></h2>
<p>If, by any reason, you need to remove coexpression pair analyzes, all you need to do is pass
the filename used to load the analyzes to the remove_relationships command: remove_relationship –file &lt;coexpression_file&gt;’
<strong>Every coexpression relations from coexpression_file (e.g. ‘pcc.mcl.txt’ from LSTrAP) will be deleted on cascade</strong>.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>manage.py<span class="w"> </span>remove_relationship<span class="w"> </span>--help
</pre></div>
</div>
<table class="docutils align-default">
<tbody>
<tr class="row-odd"><td><p>–file</p></td>
<td><p>‘mcl.clusters.txt’ file *</p></td>
</tr>
</tbody>
</table>
<p>* example output file from LSTrAP software.</p>
</section>
<section id="load-coexpression-clusters">
<h2><span class="section-number">2.12.3. </span>Load Coexpression clusters<a class="headerlink" href="#load-coexpression-clusters" title="Link to this heading"></a></h2>
<p>Other type of coexpression analyzes involve clustering features using its correlation values.
LSTrAP does that using the ‘mcl’ software. To load data from such analyzes, the input file must be
headless and fields tab separated, with each line representing one cluster, and each column representing
one gene/feature from that cluster. The first column should represent the cluster name and must have the
format: “&lt;cluster_name&gt;:”.
Three-cluster sample from such a file is shown below, the first line
represents a cluster with three individuals, the second line a cluster with two, and the third line an
orphaned cluster with only one individual (obs: orphaned clusters are discarded).:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>ath_1:<span class="w">    </span>AT3G18715.1.TAIR10<span class="w">      </span>AT3G08790.1.TAIR10<span class="w">      </span>AT5G42230.1.TAIR10
ath_2:<span class="w">    </span>AT1G27040.1.TAIR10<span class="w">      </span>AT1G71692.1.TAIR10
ath_3:<span class="w">    </span>AT5G24750.1.TAIR10
</pre></div>
</div>
<p>Note: The genes/features from each column must be loaded previously.</p>
<p>To load such a file, one can run the command below:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>manage.py<span class="w"> </span>load_coexpression_clusters<span class="w"> </span>--file<span class="w"> </span>mcl.clusters.txt<span class="w"> </span>--organism<span class="w"> </span><span class="s1">&#39;Arabidopsis thaliana&#39;</span>
</pre></div>
</div>
<ul class="simple">
<li><p>Loading this file can be faster if you increase the number of threads (–cpu).</p></li>
</ul>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>manage.py<span class="w"> </span>load_coexpression_clusters<span class="w"> </span>--help
</pre></div>
</div>
<table class="docutils align-default">
<tbody>
<tr class="row-odd"><td><p>–file</p></td>
<td><p>‘mcl.clusters.txt’ file *</p></td>
</tr>
<tr class="row-even"><td><p>–organism</p></td>
<td><p>Scientific name (e.g.: ‘Arabidopsis thaliana’)</p></td>
</tr>
<tr class="row-odd"><td><p>–soterm</p></td>
<td><p>‘mRNA’ sequence ontology term <em>default</em></p></td>
</tr>
<tr class="row-even"><td><p>–cpu</p></td>
<td><p>Number of threads</p></td>
</tr>
</tbody>
</table>
<p>* example output file from LSTrAP software.</p>
</section>
<section id="remove-coexpression-clusters">
<h2><span class="section-number">2.12.4. </span>Remove coexpression clusters<a class="headerlink" href="#remove-coexpression-clusters" title="Link to this heading"></a></h2>
<p>If, by any reason, you need to remove coexpression cluster analyzes, you need to pass
the controlled vocabulary term ‘coexpression group’ and the organism scientific name to
the command remove_feature_group: <em>remove_feature_group –cvterm ‘coexpression group’ –organism ‘Arabidopsis thaliana’</em> *</p>
<p>*Every coexpression group relations from that organism will be deleted on cascade**.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>manage.py<span class="w"> </span>remove_feature_annotation<span class="w"> </span>--help
</pre></div>
</div>
<table class="docutils align-default">
<tbody>
<tr class="row-odd"><td><p>–cvterm TERM</p></td>
<td><p>mandatory: ‘coexpression group’</p></td>
</tr>
<tr class="row-even"><td><p>–organism ORGANISM</p></td>
<td><p>Scientific name (e.g.: ‘Oryza sativa’)</p></td>
</tr>
</tbody>
</table>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="load_rnaseq.html" class="btn btn-neutral float-left" title="2.11. Loading RNA-seq data" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="load_vcf.html" class="btn btn-neutral float-right" title="2.13. Loading VCF files" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2018, Embrapa.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>